src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs

```diff
@@ -48,9 +54,19 @@ internal T[] Rent()
                 // Check if all of our buffers have been rented
                 if (_index < _data.Length)
                 {
-                    buffer = _data[_index] ?? new T[_bufferLength];
+                    buffer = _data[_index];
+                    if (buffer == null)
+                    {
+                        buffer = new T[_bufferLength];
+                        _eventSource.BufferCreated(buffer.GetHashCode(), _bufferLength, GetHashCode());
+                    }
                     _data[_index++] = null;
                 }
+                else if (_exhaustedEventSent == false)
+                {
+                    _eventSource.BucketExhausted(GetHashCode(), _bufferLength, _data.Length);
+                    _exhaustedEventSent = true;
+                }
```

Comment:
Could you add some XML comments to the event methods to help clarify what they are and when they're fired?

FWIW, I find the BufferCreated vs BufferAllocated naming confusing; I think of Created and Allocated interchangeably in this context.  Are you saying the difference is whether we've already created N buffers associated with a bucket: if this is less than or equal to the Nth, Created is fired, and if it's greater than the Nth, Allocated is fired?  If so, should they be the same event with an additional argument that provides whatever additional context you're trying to convey?